### PR TITLE
[Shell] Show an alert in StatusMonitor if the Workbench is not visible

### PR DESCRIPTION
**Pull request by jsuarezruiz**
Having the Get To Code Dialog and not having the Workbench visible, can not see the StatusView monitor feedback.
Added changes to use an alert instead of using the monitor if the workbench is not visible.

Fixes VSTS #896537

### CODE REVIEW

**Comment by slluis** [comment]
If the workbench is not visible, we should show the progress somewhere else, for example using a MessageDialogProgressMonitor.

**Comment by sevoku** [comment]
> If the workbench is not visible, we should show the progress somewhere else, for example using a MessageDialogProgressMonitor.

We should probably create an aggregated monitor with MessageDialogProgressMonitor and StatusMonitor instead of the simple StatusMonitor, if the workbench is hidden.

**Comment by jsuarezruiz** [comment]
We was using an `AggregatedMonitor` with the `StatusProgressMonitor` and `MessageDialogProgressMonitor`. So, the only changes are related to the **dispose** of the MessageMonitor depending  by the Workbench visibility; along with a fix in the error text in MessageMonitor (we mistakenly made a ToString of ProgressError).

File: main/src/addins/VersionControl/MonoDevelop.VersionControl/MonoDevelop.VersionControl/CheckoutCommand.cs [hunk 1]
```diff
@@ -49,7 +49,7 @@ protected override ProgressMonitor CreateProgressMonitor ()
 			{
 				return new MonoDevelop.Core.ProgressMonitoring.AggregatedProgressMonitor (
 					base.CreateProgressMonitor (),
-					new MonoDevelop.Ide.ProgressMonitoring.MessageDialogProgressMonitor (true, true, true, true)
+					new MonoDevelop.Ide.ProgressMonitoring.MessageDialogProgressMonitor (true, false, true, true)
```
Comment (sevoku) [inline main/src/addins/VersionControl/MonoDevelop.VersionControl/MonoDevelop.VersionControl/CheckoutCommand.cs]: why disallow cancellation?

File: main/src/addins/VersionControl/MonoDevelop.VersionControl/MonoDevelop.VersionControl/VersionControlTask.cs [hunk 2]
```diff
@@ -63,14 +61,15 @@ void BackgroundWorker ()
 						tracker.ReportError (msg, exception);
 					}
 				}
-				Wakeup ();
+				Wakeup (IsWorkbenchVisible());
 			}, Runtime.MainTaskScheduler);
 		}
 
-		public void Wakeup() {
+		public void Wakeup(bool dispose) {
```
Comment (sevoku) [inline main/src/addins/VersionControl/MonoDevelop.VersionControl/MonoDevelop.VersionControl/VersionControlTask.cs]: Do we need this API change? I think Wakeup should just take care of this.

File: main/src/addins/VersionControl/MonoDevelop.VersionControl/MonoDevelop.VersionControl/VersionControlTask.cs [hunk 3]
```diff
@@ -63,14 +61,15 @@ void BackgroundWorker ()
 						tracker.ReportError (msg, exception);
 					}
 				}
-				Wakeup ();
+				Wakeup (IsWorkbenchVisible());
```
Comment (sevoku) [inline main/src/addins/VersionControl/MonoDevelop.VersionControl/MonoDevelop.VersionControl/VersionControlTask.cs]: does this need to be a method? Why not just `IdeApp.Workbench.RootWindow?.Visible == true`?

File: main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Gui/StatusProgressMonitor.cs [hunk 4]
```diff
@@ -34,7 +34,7 @@
 
 namespace MonoDevelop.Ide.Gui
 {
-	internal class StatusProgressMonitor: ProgressMonitor
+	internal class StatusProgressMonitor : ProgressMonitor
```
[the remark this answers is not in the document]
Comment (jsuarezruiz) [reply main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Gui/StatusProgressMonitor.cs]: Yes, only formatting changes. I can recover the original format.

File: main/src/addins/VersionControl/MonoDevelop.VersionControl/MonoDevelop.VersionControl/CheckoutCommand.cs (hunk 1 above)
Comment (jsuarezruiz) [reply main/src/addins/VersionControl/MonoDevelop.VersionControl/MonoDevelop.VersionControl/CheckoutCommand.cs]: Recovered the original state.

File: main/src/addins/VersionControl/MonoDevelop.VersionControl/MonoDevelop.VersionControl/VersionControlTask.cs (hunk 3 above)
Comment (jsuarezruiz) [reply main/src/addins/VersionControl/MonoDevelop.VersionControl/MonoDevelop.VersionControl/VersionControlTask.cs]: Not necessary (we only use the method here). Applied your suggestion.

File: main/src/addins/VersionControl/MonoDevelop.VersionControl/MonoDevelop.VersionControl/VersionControlTask.cs [hunk 5]
```diff
@@ -70,7 +68,8 @@ void BackgroundWorker ()
 		public void Wakeup() {
 			try {
 				tracker.EndTask();
-				tracker.Dispose ();
+				if(IdeApp.Workbench.RootWindow?.Visible == true)
+					tracker.Dispose ();
```
Comment (sevoku) [inline main/src/addins/VersionControl/MonoDevelop.VersionControl/MonoDevelop.VersionControl/VersionControlTask.cs]: I'm not sure if it's really ok just not to dispose here. @Therzok ?

Maybe we could move the whole code to `Finished()` and let the subclasses decide? This would allow i.e. the `CheckoutCommand` to conditionally dispose the `MessageDialogProgressMonitor` if needed and just dispose others.

Comment (Therzok) [reply main/src/addins/VersionControl/MonoDevelop.VersionControl/MonoDevelop.VersionControl/VersionControlTask.cs]: @sevoku if a command needs a status monitor to outlive the operation, we should let the command implementation decide

Comment (jsuarezruiz) [reply main/src/addins/VersionControl/MonoDevelop.VersionControl/MonoDevelop.VersionControl/VersionControlTask.cs]: I have applied changes to keep the Dispose logic as it was. Now we show a dialog with error information if the Workbench is not visible.

File: main/src/addins/VersionControl/MonoDevelop.VersionControl/MonoDevelop.VersionControl/VersionControlTask.cs [hunk 6]
```diff
@@ -83,5 +82,12 @@ void BackgroundWorker ()
 		protected void Warn(string logtext) {
 			tracker.ReportWarning(logtext);
 		}
+
+		void ReportError (string message, Exception exception)
+		{
+			tracker.ReportError (message, exception);
```
Comment (sevoku) [inline main/src/addins/VersionControl/MonoDevelop.VersionControl/MonoDevelop.VersionControl/VersionControlTask.cs]: Probably we should always prepend "Version Control operation failed: " and remove that from the message in the caller.

File: main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Gui/StatusProgressMonitor.cs [hunk 7]
```diff
@@ -34,7 +34,7 @@
 
 namespace MonoDevelop.Ide.Gui
 {
-	internal class StatusProgressMonitor: ProgressMonitor
```
Comment (sevoku) [inline main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Gui/StatusProgressMonitor.cs]: unrelated formatting changes

File: main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Gui/StatusProgressMonitor.cs [hunk 8]
```diff
@@ -44,8 +44,8 @@ internal class StatusProgressMonitor: ProgressMonitor
 		string title;
 		StatusBarContext statusBar;
 		Pad statusSourcePad;
-
```
Comment (sevoku) [inline main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Gui/StatusProgressMonitor.cs]: unrelated formatting changes

File: main/src/addins/VersionControl/MonoDevelop.VersionControl/MonoDevelop.VersionControl/VersionControlTask.cs (hunk 6 above)
Comment (jsuarezruiz) [reply main/src/addins/VersionControl/MonoDevelop.VersionControl/MonoDevelop.VersionControl/VersionControlTask.cs]: Changed.

File: main/src/addins/VersionControl/MonoDevelop.VersionControl/MonoDevelop.VersionControl/VersionControlTask.cs [hunk 9]
```diff
@@ -83,5 +82,12 @@ void BackgroundWorker ()
...
+			if (IdeApp.Workbench.RootWindow?.Visible == false && exception != null)
```
[the remark this answers is not in the document]
Comment (jsuarezruiz) [reply main/src/addins/VersionControl/MonoDevelop.VersionControl/MonoDevelop.VersionControl/VersionControlTask.cs]: Not really. Applied changes to exception show a message anyway.

**Comment by sevoku** [comment]
@slluis this change is now reduced to VCS only, we need to do much more to fix and improve the MessageDialogProgressMonitor which is out of scope here.

**Comment by sevoku** [comment]
@monojenkins backport release-8.2
